File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/composite/helpers/node_builder.dart
import '../types/json.dart';
import '../component.dart';
import '../primitive_node.dart';
import '../array_composite.dart';
import '../object_composite.dart';

JsonComponent buildNode(String keyLabel, Json value) {
  if (value is List) return ArrayNode(keyLabel, value, buildNode);
  if (value is Map<String, dynamic>) return ObjectNode(keyLabel, value, buildNode);
  return PrimitiveNode(keyLabel, value);
}

File: implementations/react-native/design-patterns-rn/src/patterns/structural/composite/helpers/node-builder.ts
import { ArrayNode } from '../array-composite';
import { ObjectNode } from '../object-composite';
import { PrimitiveNode } from '../leaf';
import type { JsonValue as Json } from '../types/json';
import type { JsonComponent } from '../component';

export function buildNode(keyLabel: string, json: Json): JsonComponent {
  if (json === null) return new PrimitiveNode(keyLabel, null);
  if (Array.isArray(json)) {
    const children = json.map((v, i) => buildNode(String(i), v));
    return new ArrayNode(keyLabel, children);
  }
  if (typeof json === 'object') {
    const children = Object.keys(json).map(k => buildNode(k, (json as any)[k]));
    return new ObjectNode(keyLabel, children);
  }
  // primitive
  return new PrimitiveNode(keyLabel, json as string | number | boolean);
}
